src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentProjectsFinder.cs

```diff
@@ -276,28 +282,49 @@ private static bool HasReferenceToAssembly(Project project, string assemblyName,
         {
             Contract.ThrowIfFalse(project.SupportsCompilation);
 
-            if (!project.TryGetCompilation(out var compilation))
-            {
-                // WORKAROUND:
-                // perf check metadata reference using newly created empty compilation with only metadata references.
-                compilation = project.Services.GetRequiredService<ICompilationFactoryService>().CreateCompilation(
-                    project.AssemblyName, project.CompilationOptions!);
-
-                compilation = compilation.AddReferences(project.MetadataReferences);
-            }
+            // Defer creating compilation if we can avoid it.
+            Compilation? compilation = null;
 
             foreach (var reference in project.MetadataReferences)
             {
                 cancellationToken.ThrowIfCancellationRequested();
 
-                if (compilation.GetAssemblyOrModuleSymbol(reference) is IAssemblySymbol symbol &&
-                    symbol.Name == assemblyName)
+                if (reference is not PortableExecutableReference peReference)
+                    continue;
+
+                var metadataId = SymbolTreeInfo.GetMetadataIdNoThrow(peReference);
+                if (metadataId is null)
+                    continue;
+
+                if (!s_metadataIdToAssemblyName.TryGetValue(metadataId, out var name))
```

Comment:
>if (!s_metadataIdToAssemblyName.TryGetValue(metadataId, out var name)) [](http://example.com/codeflow?start=16&length=70)

Would it make sense to do a pass through project.MetadataReferences and see if there are any cached entries matching the name. If not, then do the pass through where the compilations can be created?

Reply:
yes.